File: solutions/377.combination-sum-iv/combination-sum-iv.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
// const combinationSum4 = (nums, target) => {
//   if(nums.length === 0) return 0
//   let map = new Map()

//   const dfs = (left) => {
//     if(left === 0) return 1
//     if(left<0) return 0
//     if(map.has(left)) return map.get(left)

//     let count = 0
//     for(let i = 0; i<nums.length; i++) {
//       count+= dfs(left - nums[i], map)
//     }
//     map.set(left, count)
//     return count
//   }

//   return dfs(target, map)
// }

// const combinationSum4 = (nums, target) => {
//   if (target === 0) return 1

//   let res = 0
//   for (let i = 0; i < nums.length; i++) {
//     if (target >= nums[i]) {
//       res += combinationSum4(nums, target - nums[i])
//     }
//   }
//   return res
// }

const combinationSum4 = function(nums, target) {
  let dp = [1]
  // i -> target
  // dp[i] -> number of combinations

  for (let i = 1; i <= target; i++) {
    dp[i] = 0
    for (let j = 0; j < nums.length; j++) {
      let item = nums[j]
      if (i - item < 0) continue

      dp[i] += dp[i - item]
    }
  }

  return dp[target]
}

File: solutions/377.combination-sum-iv/combination-sum-iv.py
from collections import deque


class Solution:
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        memo = {}

        def dfs(left):
            if left in memo:
                return memo[left]

            cnt = 0
            for n in nums:
                if n > left:
                    break

                if n == left:
                    cnt += 1
                    break

                cnt += dfs(left - n)
            memo[left] = cnt
            return cnt

        return dfs(target)
